docs: create decision record about the release process (#116)
eclipse-tractusx/tractusx-edc: docs/development/decision-records/2023-02-09-release-process/README.md

  currently released version. Nothing else.
- feature development happens _in developers' forks only_ to keep the Git reflog of the `origin` clean.

## Rationale

Having releases depend on snapshot versions of upstream projects, such as EDC, is inherently dangerous, particularly
when that dependency has not yet reached a final state and breaking changes are to be expected. Most problems will stem
from breaking changes, such as Java SPIs, APIs and changes in service contracts.

Up until now, the only way out was cherry-picking, which is extremely cumbersome and error-prone, and requires a
parallel build pipeline to publish the cherry-picked artifacts of EDC (and potentially others). With the approach
presented here, cherry-picking is still an option, but there are easier alternatives to it. 

Every release version published by tractusx-edc must be reproducible at any time.

## Approach

### Use EDC `SNAPSHOT` versions during development

During feature development we only use `-SNAPSHOT` versions of EDC packages. It is assumed that when the build breaks
due to changes in upstream, the fix can be done quickly and easily, much more so than working off technical

> debt that would otherwise accumulate over several months. Builds on `develop` are therefore _not repeatable_, but that

Author: the problem with Snapshots is that they could get overwritten or deleted. 
That means, that EDC's `0.0.1-SNAPSHOT` versions will get overwritten on each EDC build. Also, when OSSRH decides to clean out their storage, and one of our builds depends on a now-deleted "named" snapshot (e.g. `0.0.1-20230315-SNAPSHOT`), we're screwed.

However, during feature development, that is not an issue, because `develop` will depend on EDC `X.Y.Z-SNAPSHOT`, which is always there, but it may be different from day to day.

> debt that would otherwise accumulate over several months. Builds on `develop` are therefore _not repeatable_, but that

Reviewer: Thanks for the explanation! That makes things more clear for me!

downside is easily offset by the tighter alignment with and smaller technical debt and integration pain with the
upstream EDC.

> ### Use release versions of EDC in releases

Author: I was referring to "releases" in a Maven-sense: everything that has the `-SNAPSHOT` suffix can be overwritten or deleted at any time. Consequentially, `0.0.1-xyz-SNAPSHOT` versions of EDC are not guaranteed to exist forever (on the OSSRH Snapshot repo).

Every artifact that *does not have** that suffix, is considered a "release". While snapshots only go to the OSSRH Snapshot repository, "Releases" get synched to MavenCentral, so they are permanent and cannot be overwritten.

This does **not** imply a certain release schedule, theoretically you could produce a "release" every day

- Creating branches: https://github.com/marketplace/actions/create-branch
- Creating tags using GitHub's
  API: https://github.com/eclipse-edc/Connector/blob/b24a5cacbc9fcabdfd8020d779399b3e56856661/.github/workflows/release-edc.yml#L21 (
  example)

Once a release is created, the EDC upstream version must not change anymore, unless there is good reason to do so, for
example, a defect, that needs to be fixed upstream. At that point a decision can also be made to employ a cherry-pick model, in case the
upstream's development has progressed to the point of breaking changes.

### Changes to the branching model

Tractusx-edc's branching model is already very close to
the [GitFlow branching model](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow#:~:text=What%20is%20Gitflow%3F,lived%20branches%20and%20larger%20commits)
which is good. The following changes need to be made:

- feature development happens in forks only. Thus, `feature/` branches don't exist in the `origin`. This is important
  when moving to OSS.
- adhere to GitFlow branch naming conventions, i.e. no `docu`, `tryout_something` branches. Increases readability and

>   clarity, improves tool support.

Author: exactly. people can name their branches however they want. Somewhat related, I would look into enforcing Conventional Commit messages.

   commits are made on `hotfix/0.3.1`. This is needed when cherry-picking is not available due to incompatibilities
   between `develop` and the hotfix branch due to intermittent changes.
3. The fix is only relevant for the `0.3.1` hotfix, it is not needed in `develop`. This can happen, when the problem is
   not present on `develop`, because it was already implicitly fixed, or otherwise doesn't exist.

This might produce many branches, and the first `hotfix` makes the release obsolete, but it will greatly help
readability and make a release's history readily apparent.

### Nightly builds

Nightly builds are generated according to a fixed schedule (as suggested
in [this Jira issue](https://jira.catena-x.net/browse/A1IDSC-408)). Upstream EDC will soon begin to publish nightly
builds as actual releases (as opposed to: snapshots) to a separate OSSRH-operated repository (see
[this EDC decision record](https://github.com/eclipse-edc/Connector/tree/main/docs/developer/decision-records/2023-02-10-nightly-builds)).

Unfortunately there is no way to automatically trigger the tractusx-edc build whenever a new EDC nightly is

> created. The most reliable method is to periodically query for the latest EDC nightly, e.g. leveraging GitHub's `dependabot` feature, or using the method highlighted in the aforementioned Jira issue:

Reviewer: ```suggestion
created. The most reliable method is to periodically query for the latest EDC nightly, e.g. leveraging GitHub's
`dependabot` feature, or using the following `curl` command:
```

```shell
curl <OSSHR_RELEASES_REPO_URL>/org/eclipse/edc/connector-core/maven-metadata.xml | xmllint --xpath "//metadata/versioning/versions/version[last()]" -
```

That would return something like to `0.0.1-20230213`. As soon as the version string (here: Feb 13th, 2023) matches the
current date, we can start the nightly. If the EDC nightly doesn't appear within a set timeframe, we throw an error.

## Notes for becoming OpenSource

- All artifacts (docker images, helm charts, Maven artifacts) should be published to well-known and publicly accessible
  locations such as MavenCentral, DockerHub, etc. The GitHub Packages repository is only accessible to authenticated

>   users.

Reviewer: Does this also include our previous releases?

Author: Personally I don't see much value in that, provided that the previous releases stay in place.

Reviewer: Yes, some opinion.

> - When and if Tractus-X becomes an Eclipse project, we'll have to adopt the Eclipse Foundation's publishing guidelines,

Reviewer: ```suggestion
- When the project was migrated to be an Eclipse project, we'll have to adopt the Eclipse Foundation's publishing guidelines,
```
